### PR TITLE
Migrate to leveraging the tokio-io crate

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit updates to the most recent versions (released today) of the various
Tokio libraries in use. Namely the `tokio_core::io` module has now been
deprecated in favor of an external `tokio-io` crate. This commit pulls in that
crate and uses the `AsyncRead + AsyncWrite` abstraction instead of `Io` from
tokio-core.

- [ ] The commit messages match the guidelines in https://github.com/hyperium/hyper/blob/master/CONTRIBUTING.md#git-commit-guidelines

### CODE REVIEW

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: 0dcfd18e51b432f2b20a6345638577fc4256922c
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/hyperium/hyper/blob/master/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by alexcrichton** [comment]
Ah and I should also mention that I dropped the `relay` crate while I was at it in favor of `futures::unsync`, but I can back out that change if you'd prefer.

File: src/http/conn.rs [hunk 1]
```diff
@@ -155,7 +155,8 @@ where I: Io,
     }
 
     fn maybe_park_read(&mut self) {
-        if self.io.poll_read().is_ready() {
+        // TODO: this needs fixing
```
Comment (seanmonstar) [inline src/http/conn.rs]: This was part of trying to keep from busy looping, how broken is the "this needs fixing" part :D

Comment (alexcrichton) [inline src/http/conn.rs]: d'oh right, sorry I thought I had handled this elsewhere already.

I was wondering if you know how the best way might be to handle this off-hand?

Comment (seanmonstar) [inline src/http/conn.rs]: Hrm, so as the comment below tries to describe, this situation would occur when hyper has read the end of a body, such as described by a `Content-Length` header, and then stops, as it's waiting on the response side to determine if it should read the next request.

If the socket were to already have some of the bytes in its buffer, then our not reading to end would mean `TcpStream` would not unpark the task again to read more. That's a scenario when hyper needs to park, and then unpark later when the response finishes and keeps the connection alive.

So, to get the same as poll_read from now on, does that mean we need to store our own `read_has_blocked` flag?

Comment (alexcrichton) [inline src/http/conn.rs]: Yeah I believe so, I'll look into adding such a flag.

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: 41fc8dff3360a955693247eee416ba0c5186f5ee
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/hyperium/hyper/blob/master/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by alexcrichton** [comment]
Ok I believe that should handled the blocked read case?

**Comment by seanmonstar** [comment]
Great, thanks!

**Pull request merged**